VoxelSpace.html

```diff
@@ -404,9 +439,43 @@
         var invz = 1. / z * 240.;
         for(var i=0; i<screenwidth|0; i=i+1|0)
         {
-            var mapoffset = ((Math.floor(ply) & mapwidthperiod) << map.shift) + (Math.floor(plx) & mapheightperiod)|0;
-            var heightonscreen = (camera.height - map.altitude[mapoffset]) * invz + camera.horizon|0;
-            DrawVerticalLine(i, heightonscreen|0, hiddeny[i], map.color[mapoffset]);
+            var final_color;
+            var altitude;
+            if(bilinear)
+            {
+                var plxt = plx - Math.floor(plx);
+                var plyt = ply - Math.floor(ply);
+
+                var plx_left = (Math.floor(plx) & mapwidthperiod);
+                var plx_right = (Math.floor(plx+1) & mapwidthperiod);
+                var ply_lower = (Math.floor(ply) & mapwidthperiod) << map.shift;
+                var ply_upper = (Math.floor(ply+1) & mapwidthperiod) << map.shift;
+
+                var llindex = ply_lower + plx_left;
+                var ulindex = ply_upper + plx_left;
+                var urindex = ply_upper + plx_right;
+                var lrindex = ply_lower + plx_right;
+
+                // Interpolate the top and bottom edges to a point on each, then interpolate
+                // between those to get the final color.
+                var upper_color = ColorInterp(map.color[ulindex], map.color[urindex], plxt);
+                var lower_color = ColorInterp(map.color[llindex], map.color[lrindex], plxt);
+                final_color = ColorInterp(lower_color, upper_color, plyt);
```

Comment:
Perf improvement opportunity (probably minor): decompose the 4 corners from uint32's into RGB values ahead of time (better yet, make map.color already store separate R, G, and B values), do the bilinear filtering, and then recompose only at the end. This would save recomposing **upper_color** and **lower_color** into uint32's, then immediately decomposing them again.